src/Microsoft.ML.Data/Transforms/DropSlotsTransform.cs

```diff
@@ -331,40 +365,140 @@ private void GetSlotsMinMax(Column col, out int[] slotsMin, out int[] slotsMax)
+                {
+                    if (slotsMin[i][j] <= slotsMax[i][iDst] + 1)
+                        slotsMax[i][iDst] = Math.Max(slotsMax[i][iDst], slotsMax[i][j]);
+                    else
+                    {
+                        iDst++;
+                        slotsMin[i][iDst] = slotsMin[i][j];
+                        slotsMax[i][iDst] = slotsMax[i][j];
+                    }
+                }
+                iDst++;
+                Array.Resize(ref slotsMin[i], iDst);
+                Array.Resize(ref slotsMax[i], iDst);
+            }
+        }
+
+        private static (string input, string output)[] GetColumnPairs(ColumnInfo[] columns)
+            => columns.Select(c => (c.Input, c.Output ?? c.Input)).ToArray();
+
+        private static bool AreRangesValid(int[][] slotsMin, int[][] slotsMax)
+        {
+            if (slotsMin.Length != slotsMax.Length)
+                return false;
+            for (int iinfo = 0; iinfo < slotsMin.Length; iinfo++)
+            {
+                var prevmax = -2;
+                for (int i = 0; i < slotsMin[iinfo].Length; i++)
+                {
+                    if (!(0 <= slotsMin[iinfo][i] && slotsMin[iinfo][i] < int.MaxValue))
+                        return false;
+                    if (!(0 <= slotsMax[iinfo][i] && slotsMax[iinfo][i] < int.MaxValue))
+                        return false;
+                    if (!(slotsMin[iinfo][i] <= slotsMax[iinfo][i]))
+                        return false;
+                    if (!(slotsMin[iinfo][i] - 1 > prevmax))
+                        return false;
+                    prevmax = slotsMax[iinfo][i];
+                }
+            }
+            return true;
+        }
+
+        protected override IRowMapper MakeRowMapper(Schema schema)
+            => new Mapper(this, schema);
+
+        private sealed class Mapper : OneToOneMapperBase
         {
-            Contracts.AssertValue(slotDropper);
-            Contracts.AssertValue(input);
-            Contracts.AssertNonEmpty(slotsMin);
-            Contracts.AssertNonEmpty(slotsMax);
-            Contracts.Assert(slotsMin.Length == slotsMax.Length);
-            Contracts.Assert(0 <= iinfo && iinfo < Infos.Length);
-
-            categoricalRanges = null;
-            // Register for metadata. Propagate the IsNormalized metadata.
-            using (var bldr = Metadata.BuildMetadata(iinfo, input, Infos[iinfo].Source,
-                MetadataUtils.Kinds.IsNormalized, MetadataUtils.Kinds.KeyValues))
-            {
-                var typeSrc = Infos[iinfo].TypeSrc;
+            private readonly SlotsDroppingTransformer _parent;
+            private readonly int[] _cols;
+            private readonly ColumnType[] _srcTypes;
+            private readonly ColumnType[] _dstTypes;
+            private readonly SlotDropper[] _slotDropper;
+            // Track if all the slots of the column are to be dropped.
+            private readonly bool[] _suppressed;
```

Comment:
why it's one dimensional array? don't we have array of columns and in each column we have slots?

Reply:
The ith element of this array is true if all slots are dropped from the ith column. We are given a set of columns and related ranges of slots to drop for each column. This indicates if for the ith column all the slots will be dropped. So this does not indicate if a give slot is dropped (that's what the ranges are for), but rather whether all slots will be dropped.

---
In reply to: [236089331](https://github.com/dotnet/machinelearning/pull/1683#discussion_r236089331) [](ancestors = 236089331)